alacritty/src/renderer/rect.rs

```diff
@@ -0,0 +1,259 @@
+use std::mem::size_of;
+use std::ptr;
+
+use alacritty_terminal::term::SizeInfo;
+
+use crate::gl;
+use crate::gl::types::*;
+use crate::renderer::rects::RenderRect;
+use crate::renderer::{create_program, create_shader, Error, ShaderCreationError};
+
+/// Maxmimum number of rect vertices per batch. Limited by 16-bit integer indices.
+const MAX_VERTICES: usize = u16::max_value() as usize + 1;
+
+/// Shader sources for rect rendering program.
+static RECT_SHADER_F_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.f.glsl");
+static RECT_SHADER_V_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.v.glsl");
+static RECT_SHADER_F: &str = include_str!("../../res/rect.f.glsl");
+static RECT_SHADER_V: &str = include_str!("../../res/rect.v.glsl");
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Rgba {
+    r: u8,
+    g: u8,
+    b: u8,
+    a: u8,
+}
+
+/// Vertex to store solid-color rectangle data.
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Vertex {
+    // Normalized screen coordinates.
+    // TODO these can certainly be i16.
+    x: f32,
+    y: f32,
+
+    // Color.
+    color: Rgba,
+}
+
+/// Structure to store and group together rect-related vertices data and GL objects.
+#[derive(Debug)]
+pub struct RectRenderer {
+    // GL buffer objects. VAO stores attribute and ebo bindings.
+    vao: GLuint,
+    vbo: GLuint,
+    ebo: GLuint,
+
+    program: RectShaderProgram,
+
+    // Accumulated vertices data. Allocated space reused between frames.
+    vertices: Vec<Vertex>,
```

Comment:
This is getting kind of philosophical :).
Defining a Vec where you need it is only simple if you don't care about externalities, which is introduction of allocator activity, which is rather complex. It may not show up in regular profiling, but sometime somewhere in one frame out of 1000 you'd get a spike.
This is one of the reasons why many mature game engines preallocate pools for everything, have custom container implementations, and have "no allocations in renderloop" policy.
Weeding out allocations from codebase that wasn't allocator-aware previously (or even discovering that you have this death-by-1000-cuts problem) is often not trivial.

Here the worst case mem cost of that Vec is 800Kb (+800Kb in GL buffer), and even if it is unused and memory pressure scenario arises, the OS will happily swap it out. This worst-case wasted space can be lowered by having lower value for MAX_VERTICES.